### DISCUSSION TITLE
How can we manage copyright and access to 3D models?

### DISCUSSION THREAD

**Discussion by andrewtsymbal** [google/model-viewer#1886] [Q&A]
### Description

Model-viewer is a great idea! But there is a serious issue with the 3D files protection.
Because anyone can open the code and download source .glb and .usdz 3D files.

For example if I have a web store and sell models for 3D printing. And I want my customers to be able to play with the model and check it from all the sides before purchasing.
I can do it with the model-viewer. But at the same time, 3D files that I'll use for preview are not protected from downloading. 

Moreover, if it's an image I can add a watermark on it to protect from stealing. And if someone will be using my image I can send an abuse report to google or web hoster. But if someone stole my 3D files and will be using them on another website, how can I send an abuse report?

An example Burberry Arthur sneakers - you can see a 3D file only in the google search.
But it's not appearing on the product page. And in this case, how can I send a report if someone uses my 3D files in the same way?

#### An Example

https://uk.burberry.com/beta/nylonsuede-arthur-sneakers-p80210361
https://i.imgur.com/ixNcEfJ.png

**Comment by cdata** [comment]
@andrewtsymbal thanks for getting in touch with us. You are absolutely right that this is a tricky problem, and the solution is going to depend a lot on your specific use case.

Before I reply directly to your questions, I would like to make an observation:

glTF and USDZ are considered delivery formats; they are the format of a 3D file that is a by-product of exporting from a design tool. They are not the source files. They are more like JPEG/PNG, which are output formats that may be produced from RAW/PSD/AI source files.

> Because anyone can open the code and download source .glb and .usdz 3D files.

This is true, and also applies to most things that you publish to the web today. Just like other media formats (video and images) on the web, the ability to download the content is required in order for a browser to present the content to the user.

Now, with video there is a special carve out for [DRM](https://developers.google.com/web/updates/2017/09/chrome-62-media-updates#widevine_l1) (folks will have strong opinions about this one way or another). There is probably a case to be made for browsers to include special APIs that enable DRM on 3D models. After all, it is expensive to produce 3D models, and risky to make it possible for such content to be misappropriated by others.

One thing you can do today is limit the ability of folks to download assets by appending time-limited or access-limited tokens to the file URLs, and also set restrictive CORS headers on your website. These measures will stop folks from cross-loading your 3D models from other origins, and also stop folks from writing scripts to download your models automatically. However, this is far from a perfect solution if you wish to prevent someone from downloading your models at all.

> Moreover, if it's an image I can add a watermark on it to protect from stealing.

There are actually a lot of ways that you can do the same thing with 3D models. 3D models are fundamentally a combination of images and geometry. You can (and folks do) watermark the textures in the 3D model. You can also add a watermark in the form of geometry in the model (e.g., your 3D logo floating in the background; although it is arguably easier to edit this kind of watermark out).

> And if someone will be using my image I can send an abuse report to google or web hoster. But if someone stole my 3D files and will be using them on another website, how can I send an abuse report?

To be honest, I think the available course of action here would be the same in both cases. Perhaps I am missing something. Can you go into detail about how you think the scenarios would be different?

> An example Burberry Arthur sneakers - you can see a 3D file only in the google search.
But it's not appearing on the product page. And in this case, how can I send a report if someone uses my 3D files in the same way?

In this case, Google has made an association between the sneaker and the origin it was loaded from. Whether or not the sneaker is displayed to you on the origin is a choice by that content provider, but you surely can still report the abuse to Google and get them to stop serving the model. Does that sound right to you?

**Comment by andrewtsymbal** [comment]
Cristopher, thank you for the reply.

I'd like to share my point of view as a professional 3D artist.

All the time different blogs stealing my 3D renderings so I have to protect images with the watermarks and put a website address on them.
But even watermarks do not completely protect my work.
I had an experience when some people grab my work with the watermark and put their own above mine! 

That's why a copyright question is very important to me.

**glTF** and **USDZ** files are not like JPEG/PNG images.
Both formats contain the 3D model and textures.
Here is how [USDZ file looks like](https://i.imgur.com/AiH9Cfk.png) - it's an archive with the 3D scene and textures.

Let's say, someone download USDZ file and unpack it.
Even if textures are marked with the logo, very often, it's very easy to remove them with Photoshop. 
Then if color correct and change textures, pack everything back to USDZ. Will the search engine recognize the original model or will appear it as new unique content?

The same for glTF you can easily open 3d scene with the [Three.js editor](https://threejs.org/editor/)
and [export to the editable file format](https://i.imgur.com/jA3ESHt.png).
So it will be very easy to remove your 3D logo (if it's done with the geometry).

Today many brands start using this technology for their product.
[Here is an example](https://uk.burberry.com/medium-leather-tb-bag-p80103351)
And it's easy to find 3d files:
[.glb](https://creativetech.static.burberry.com/ar/quick-look/tb-bag/v1.1/tb-bag-black.glb)
[.usdz](https://creativetech.static.burberry.com/ar/quick-look/tb-bag/v1.1/tb-bag-black.usdz)

Brands will invest a lot of money in 3D content, do you think they will be happy that anyone can get their 3D models?

**And one more fact**: once I created a 3d model one of the very famous shoe brands and posted it on my website. It was not 100% accurate and ready for production. But the brand representative reached me and demanded to delete the 3D model because they were afraid that people can download it and make 3D printed copies for sale.

**Comment by cdata** [comment]
@andrewtsymbal I'm sympathetic to the problem you are presenting, and the best I can offer you is that this emerging content format is very new and folks are still figuring out the best practices for production and distribution of 3D models.

 > But even watermarks do not completely protect my work.

Yes, and it is the same process to remove a watermark from a texture or a photo. But, photographers still put their work online. There is a balance to be struck.

> The same for glTF you can easily open 3d scene with the Three.js editor and export to the editable file format.

Technically this is true, but this is a lossy transformation. Ultimately, your source .blend file will not be able to be reconstructed from a distributed .gltf file. You can also go out of your way to flatten your geometry and bake your textures so that they are no longer representative of the hierarchy in your source assets.

> I had an experience when some people grab my work with the watermark and put their own above mine!
> That's why a copyright question is very important to me.

Ultimately, even with DRM it is possible for a suitably motivated individual to copy your work and re-distribute it. In my opinion, the best you can do is make it difficult to copy. At some point, you will cross a line where it becomes hard for your paying audience to view your work.

To offer a real-world example: it is actually not possible for me (as a paying subscriber) to watch Netflix on a lot of my devices (like the small display I have in my kitchen) because Netflix is very strict with how distribution / copyright controls work on their platform. But, nefarious people still somehow manage to copy and re-distribute Netflix's exclusive content. The irony is that if I downloaded the illegal copies, I would have greater flexibility to watch that content on the devices that I own.

> Brands will invest a lot of money in 3D content, do you think they will be happy that anyone can get their 3D models?

Yes, I do. We work with the [Khronos 3D Commerce group](https://www.khronos.org/3dcommerce/). This is a standards body consisting of major brands that is talking about doing just this.

I'm sympathetic to the problem you are presenting, and I also think this is an important topic for our users to consider. But, our project is not going to be able to solve the problems you are describing. You will have to decide what your personal comfort level is for distributing your work, and take the appropriate steps to mitigate what you think are the risks related to copyright.

**Comment by elalish** [comment]
@andrewtsymbal If the goal of your 3D content is to be a web-rendered object, then you're right that delivering that content opens it up to exploitation. It is unlikely (for technical reasons) that DRM will come to this space any time soon. 

However, if your 3D content is meant to describe other content (e.g. actual objects available for purchase, or files meant to be 3D printed or otherwise manufactured), then you have some options. In this case, we need vastly less resolution (both geometric and texture) to make a good rendering than you need for manufacturing. A potential option is simply compression (see DRACO), which can remove enough detail that the file is useless for manufacturing but still good enough for rendering. This is akin to a photographer putting low res photos online in hopes that you'll buy a full res print.

**Comment by elmau** [comment]
Could it be possible to retrieve the file from a database?

**Comment by cdata** [comment]
@elmau certainly you can serve the file however you wish.

**Comment by pedrorocha-net** [comment]
Great to see this debate here. I'm working with product designers, to allow them to showcase their products (furnitures) online, but for them it's a big deal to have someone simply downloading their 3D model and starting to copy it. It's their work and anyone could simply start to produce it.

I'll try to follow the suggestion on compressing the files. That can help, for sure.

**Comment by clochardM33** [comment]
We got around this with a 90's style file swap.
1. Don't load the model directly.
2. Put the file you want in a php variable. 
3. In the html add a call to a php routine that copies the real file (From the variable) into a placeholder with a unique name (so you can support multiple requests on same model). We went for remote ip address and current time (watch out for IPv6 addresses).
4. Load this placeholder file as the mv sec
5. MV onload call another php routine that deletes the placeholder file.
6. Add a home-based sweep up routine to catch files that have been in the placeholder folder "Too long" - there are occasionally some.
7. Pretend it's a secure system.
So there is temporarily a file that MV (anyone) can see, but unless you are really really techy by the time you get to find the url it has gone.
Point 7 is what it all hinges on.

**Reply by clochardM33** [reply]
Point 2. Should be filename not file

**Reply by clochardM33** [reply]
And point 4 should be MV src not sec

**Reply by clochardM33** [reply]
{Sigh} And point 6 should be time-based not home-based

**Reply by elalish** [reply]
I like point 7 too! Also, pretty sure there's an edit button under the ... menu :)

**Reply by emincanyuksel** [reply]
> {Sigh} And point 6 should be time-based not home-based

hi, i want to use your method but i don't quite understand can you help me or is there a sample file or something?

**Reply by clochardM33** [reply]
Hi.
I can start off trying to explain it and see how far we get with that.
It starts with something I read on the internet that "If the browser can see it then anyone can see it". It took a while to understand the implication of this (even though it's self evidently true once you realise what it was saying). It means (to me) that if your browser has access to "it" then it must be on a server somewhere that you (the browser) can get to using a normal url.
The "it" could be anything, in this case it is the glb model files, but could equally be the MV library, some fonts, images etc.
Which in turn means that if you have a url then you can "download it" through the browser by just going to that url. After all, to a first approximation, that is all the browser is doing when you load the page.

Now the implication of this as far as MV is concerned is you cannot protect the glb files because they have to be somewhere that the browser can get to them. And by protect I mean stop them being accessed or downloaded from something outside of ModelViewer (because they have to be at the end of a url somewhere).

You can sanity check this by looking at the Network page of the f12 Developer Panel.

There is undoubtedly a really clever way of making a glb file both accessible and hidden simultaneously, but I have no idea what it is, don't have the time to look into it and don't have any money to spend getting someone else to do these steps. And on to of that it would have to be compatible with MV's loading mechanism.

This took us to the scheme we worked up which is listed in the start if this message. It is really important to note that point 7 is super important: It isn't a secure system, but it's not bad.

If the aim is to stop someone downloading the glb file outside of MV then, following the tale above, you (probably) can't do it as you have to disclose the glb url in the Network Stats panel and anyone who has pressed f12 can get to that.
An alternative is to make the url useless to anything but MV, and this is what we do.
As hinted above it is a very simple idea and doesn't require login, complicated page behaviour.

When our page loads we copy the glb file from wherever it lives on the server (where the browser cannot see) into a temporary location that the browser can see. Load from this temporary address and then when MV raises the onLoad event we delete the file. That's all there is to it.

If you then go to the url that MV has just used there is nothing there. So someone looking to download the glb file can go as many time as they like, but there is nothing to download.

It doesn't take a cyber security expert to spot the many weaknesses in this approach, but we bet on the fact that most users don't know that f12 is even a thing.

The mechanism works for us because our pages are dynamically generated using php; they are not static pages. This means we already have a scheme for running code server-side and the concept of configuration files.
If you only had static pages using js then I don't see how you could do anything because you need to swap files server-side.

In a bit more detail:
The MV parameters are all in a json file. These are things such as camera-target, camera-orbit etc. One of the parameters is the name of the glb file we want to show for that page.
The files are stored somewhere outside of wwwRoot. Php can get to them there, but webServer cannot.
When the page builds (using php) there is a routine that picks up the filename then makes a copy of it in a location that webServer can access. We also give it a pseudo random name that is derived from the Clients IP address, the time at which the request came and some random numbers. This is a way of handling multiple requests to the same file, nothing more.
Once MV has loaded everything (the onLoad event) we call a js routing that calls back to the server with the please_delete_this_file request - which is php again.
And that's it.

Reread Point 7!

**Comment by siliconvalleypilot** [comment]
Don't listen to the BS that 3D data cannot be protected (https://github.com/google/model-viewer/discussions/1886#discussioncomment-300293). Look at Sketchfab. They not only protect the 3D source files from F12 inspection but also protect 3D files from being "GPU ripped".
